src/modules/adaptors/Morpho/MorphoBlue/MorphoBlueSupplyAdaptor.sol

```diff
@@ -0,0 +1,254 @@
+        if (totalBorrowAssets >= totalSupplyAssets) return 0;
+        uint256 liquidSupply = totalSupplyAssets - totalBorrowAssets;
+        uint256 cellarSuppliedBalance = morphoBlue.expectedSupplyAssets(market, msg.sender);
+        withdrawableSupply = cellarSuppliedBalance > liquidSupply ? liquidSupply : cellarSuppliedBalance;
+    }
+
+    /**
+     * @notice Returns the cellar's balance of the supplyToken position.
+     * @param adaptorData encoded bytes32 MB id that represents the MB market for this position.
+     * @return Cellar's balance of the supplyToken position.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        Id id = abi.decode(adaptorData, (Id));
+        return _userSupplyBalance(id, msg.sender);
+
+        // below LoCs are if we use periphery library
+        // MarketParams memory market = morphoBlue.idToMarketParams(id);
+        // return _balanceOf(market); // TODO maybe reduce to just 1 param (lose the `user` param)
+    }
+
+    /**
+     * @notice Returns loanToken.
+     * @return ERC20 loanToken.
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        Id id = abi.decode(adaptorData, (Id));
+        MarketParams memory market = morphoBlue.idToMarketParams(id);
+        return ERC20(market.loanToken);
+    }
+
+    /**
+     * @notice This adaptor returns collateral, and not debt.
+     * @return Whether or not this position is a debt position
+     */
+    function isDebt() public pure override returns (bool) {
+        return false;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    /**
+     * @notice Allows strategists to lend specific asset on Morpho Blue market
+     */
+    function lendToMorphoBlue(Id _id, uint256 _assets) public {
+        _validateMBMarket(_id);
+        MarketParams memory market = morphoBlue.idToMarketParams(_id);
+        ERC20 loanToken = ERC20(market.loanToken);
+        _assets = _maxAvailable(loanToken, _assets);
+        loanToken.safeApprove(address(morphoBlue), _assets);
+        _deposit(market, _assets, address(this));
+        // Zero out approvals if necessary.
+        _revokeExternalApproval(loanToken, address(morphoBlue));
+    }
+
+    /**
+     * @notice Allows strategists to withdraw underlying asset plus interest.
+     */
+    function withdrawFromMorphoBlue(Id _id, uint256 _assets) public {
+        // Run external receiver check.
+        _externalReceiverCheck(address(this));
+        _validateMBMarket(_id);
+        MarketParams memory market = morphoBlue.idToMarketParams(_id);
+        _accrueInterest(market); // TODO - if we end up using periphery library for _balanceOf() then we may not need to kick `accrueInterest()`. We sacrifice losing some dust / noise though I think. Need to test this.
```

Comment:
I think it is best to leave the accrueInterest call in here. But I would move it down into the if statement. Cuz if the strategist is not trying to pull everything from it, then we dont need to kick it before hand cuz we know exactly how much the strategist wants to pull

Reply:
Sounds good. I've moved it into the if statement logic, and since the morphoBlue codebase should revert when trying to withdraw more than one has supplied, this should work out fine. I ran into an unexpected overflow/underflow though when adding more checks to `testWithdraw()` that I think might be worthwhile to check with Morpho team this week.